File: server/routes/people.js
const express = require('express');
const router = express.Router();
const db = require('../db/database');
const crypto = require('crypto');

// GET /api/people - List all people with optional project filter
router.get('/', (req, res) => {
  try {
    const { project_id } = req.query;
    
    let query = 'SELECT * FROM people WHERE 1=1';
    const params = [];
    
    if (project_id) {
      query += ' AND project_id = ?';
      params.push(project_id);
    }
    
    query += ' ORDER BY created_at DESC';
    
    const people = db.prepare(query).all(...params);
    res.json({ success: true, data: people });
  } catch (error) {
    console.error('Error fetching people:', error);
    res.status(500).json({ 
      success: false, 
      error: { 
        code: 'FETCH_ERROR', 
        message: 'Failed to fetch people' 
      } 
    });
  }
});

// GET /api/people/:id - Get single person
router.get('/:id', (req, res) => {
  try {
    const person = db.prepare('SELECT * FROM people WHERE id = ?').get(req.params.id);
    
    if (!person) {
      return res.status(404).json({ 
        success: false, 
        error: { 
          code: 'NOT_FOUND', 
          message: 'Person not found' 
        } 
      });
    }
    
    res.json({ success: true, data: person });
  } catch (error) {
    console.error('Error fetching person:', error);
    res.status(500).json({ 
      success: false, 
      error: { 
        code: 'FETCH_ERROR', 
        message: 'Failed to fetch person' 
      } 
    });
  }
});

// POST /api/people - Create person
router.post('/', (req, res) => {
  try {
    const { name, email, company, designation, project_id } = req.body;
    
    // Validation
    if (!name || name.trim() === '') {
      return res.status(400).json({ 
        success: false, 
        error: { 
          code: 'VALIDATION_ERROR', 
          message: 'Person name is required' 
        } 
      });
    }
    
    // Validate project_id if provided
    if (project_id) {
      const project = db.prepare('SELECT id FROM projects WHERE id = ?').get(project_id);
      if (!project) {
        return res.status(400).json({ 
          success: false, 
          error: { 
            code: 'VALIDATION_ERROR', 
            message: 'Project not found' 
          } 
        });
      }
    }
    
    const id = crypto.randomUUID();
    
    db.prepare(`
      INSERT INTO people (id, name, email, company, designation, project_id) 
      VALUES (?, ?, ?, ?, ?, ?)
    `).run(
      id, 
      name.trim(), 
      email?.trim() || null, 
      company?.trim() || null, 
      designation?.trim() || null, 
      project_id || null
    );
    
    const newPerson = db.prepare('SELECT * FROM people WHERE id = ?').get(id);
    
    res.status(201).json({ success: true, data: newPerson });
  } catch (error) {
    console.error('Error creating person:', error);
    res.status(500).json({ 
      success: false, 
      error: { 
        code: 'CREATE_ERROR', 
        message: 'Failed to create person' 
      } 
    });
  }
});

// PUT /api/people/:id - Update person
router.put('/:id', (req, res) => {
  try {
    const { id } = req.params;
    const { name, email, company, designation, project_id } = req.body;
    
    // Check if person exists
    const existingPerson = db.prepare('SELECT * FROM people WHERE id = ?').get(id);
    if (!existingPerson) {
      return res.status(404).json({ 
        success: false, 
        error: { 
          code: 'NOT_FOUND', 
          message: 'Person not found' 
        } 
      });
    }
    
    // Validate project_id if provided
    if (project_id !== undefined && project_id !== null) {
      const project = db.prepare('SELECT id FROM projects WHERE id = ?').get(project_id);
      if (!project) {
        return res.status(400).json({ 
          success: false, 
          error: { 
            code: 'VALIDATION_ERROR', 
            message: 'Project not found' 
          } 
        });
      }
    }
    
    // Update person
    db.prepare(`
      UPDATE people 
      SET name = COALESCE(?, name), 
          email = COALESCE(?, email), 
          company = COALESCE(?, company), 
          designation = COALESCE(?, designation), 
          project_id = COALESCE(?, project_id), 
          updated_at = CURRENT_TIMESTAMP 
      WHERE id = ?
    `).run(
      name !== undefined ? name.trim() : null,
      email !== undefined ? (email?.trim() || null) : null,
      company !== undefined ? (company?.trim() || null) : null,
      designation !== undefined ? (designation?.trim() || null) : null,
      project_id !== undefined ? project_id : null,
      id
    );
    
    const updatedPerson = db.prepare('SELECT * FROM people WHERE id = ?').get(id);
    
    res.json({ success: true, data: updatedPerson });
  } catch (error) {
    console.error('Error updating person:', error);
    res.status(500).json({ 
      success: false, 
      error: { 
        code: 'UPDATE_ERROR', 
        message: 'Failed to update person' 
      } 
    });
  }
});

// DELETE /api/people/:id - Delete person
router.delete('/:id', (req, res) => {
  try {
    const { id } = req.params;
    
    // Check if person exists
    const person = db.prepare('SELECT * FROM people WHERE id = ?').get(id);
    if (!person) {
      return res.status(404).json({ 
        success: false, 
        error: { 
          code: 'NOT_FOUND', 
          message: 'Person not found' 
        } 
      });
    }
    
    // Delete person (cascade will handle task_assignees)
    db.prepare('DELETE FROM people WHERE id = ?').run(id);
    
    res.json({ success: true, message: 'Person deleted successfully' });
  } catch (error) {
    console.error('Error deleting person:', error);
    res.status(500).json({ 
      success: false, 
      error: { 
        code: 'DELETE_ERROR', 
        message: 'Failed to delete person' 
      } 
    });
  }
});

module.exports = router;

File: server/internal/handlers/people.go
package handlers

import (
	"database/sql"
	"net/http"
	"strings"

	"github.com/celestask/server/internal/db"
	"github.com/celestask/server/internal/middleware"
	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

// Person represents a person in the system
type Person struct {
	ID          string  `json:"id"`
	Name        string  `json:"name"`
	Email       *string `json:"email"`
	Company     *string `json:"company"`
	Designation *string `json:"designation"`
	ProjectID   *int64  `json:"project_id"`
	CreatedAt   string  `json:"created_at"`
	UpdatedAt   string  `json:"updated_at"`
}

// CreatePersonRequest represents the request body for creating a person
type CreatePersonRequest struct {
	Name        string  `json:"name" binding:"required"`
	Email       *string `json:"email"`
	Company     *string `json:"company"`
	Designation *string `json:"designation"`
	ProjectID   *int64  `json:"project_id"`
}

// UpdatePersonRequest represents the request body for updating a person
type UpdatePersonRequest struct {
	Name        *string `json:"name"`
	Email       *string `json:"email"`
	Company     *string `json:"company"`
	Designation *string `json:"designation"`
	ProjectID   *int64  `json:"project_id"`
}

// GetPeople handles GET /api/people - List all people with optional project filter
func GetPeople(c *gin.Context) {
	databaseIface, exists := c.Get("database")
	if !exists {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Database not available"))
		return
	}

	database, ok := databaseIface.(*db.Database)
	if !ok {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Invalid database instance"))
		return
	}

	projectID := c.Query("project_id")

	var query string
	var args []interface{}

	if projectID != "" {
		query = "SELECT id, name, email, company, designation, project_id, created_at, updated_at FROM people WHERE project_id = ? ORDER BY created_at DESC"
		args = append(args, projectID)
	} else {
		query = "SELECT id, name, email, company, designation, project_id, created_at, updated_at FROM people ORDER BY created_at DESC"
	}

	rows, err := database.Query(query, args...)
	if err != nil {
		c.JSON(http.StatusInternalServerError, middleware.NewFetchError("people"))
		return
	}
	defer rows.Close()

	var people []Person
	for rows.Next() {
		var p Person
		if err := rows.Scan(&p.ID, &p.Name, &p.Email, &p.Company, &p.Designation, &p.ProjectID, &p.CreatedAt, &p.UpdatedAt); err != nil {
			c.JSON(http.StatusInternalServerError, middleware.NewFetchError("people"))
			return
		}
		people = append(people, p)
	}

	if people == nil {
		people = []Person{}
	}

	c.JSON(http.StatusOK, middleware.NewSuccessResponse(people))
}

// GetPerson handles GET /api/people/:id - Get single person
func GetPerson(c *gin.Context) {
	databaseIface, exists := c.Get("database")
	if !exists {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Database not available"))
		return
	}

	database, ok := databaseIface.(*db.Database)
	if !ok {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Invalid database instance"))
		return
	}

	id := c.Param("id")

	var p Person
	err := database.QueryRow(
		"SELECT id, name, email, company, designation, project_id, created_at, updated_at FROM people WHERE id = ?",
		id,
	).Scan(&p.ID, &p.Name, &p.Email, &p.Company, &p.Designation, &p.ProjectID, &p.CreatedAt, &p.UpdatedAt)

	if err == sql.ErrNoRows {
		c.JSON(http.StatusNotFound, middleware.NewNotFoundError("Person"))
		return
	}
	if err != nil {
		c.JSON(http.StatusInternalServerError, middleware.NewFetchError("person"))
		return
	}

	c.JSON(http.StatusOK, middleware.NewSuccessResponse(p))
}

// CreatePerson handles POST /api/people - Create new person
func CreatePerson(c *gin.Context) {
	databaseIface, exists := c.Get("database")
	if !exists {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Database not available"))
		return
	}

	database, ok := databaseIface.(*db.Database)
	if !ok {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Invalid database instance"))
		return
	}

	var req CreatePersonRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, middleware.NewValidationError("Invalid request body"))
		return
	}

	// Validate name is not empty
	req.Name = strings.TrimSpace(req.Name)
	if req.Name == "" {
		c.JSON(http.StatusBadRequest, middleware.NewValidationError("Person name is required"))
		return
	}

	// Validate project_id if provided
	if req.ProjectID != nil {
		var count int
		err := database.QueryRow("SELECT COUNT(*) FROM projects WHERE id = ?", *req.ProjectID).Scan(&count)
		if err != nil || count == 0 {
			c.JSON(http.StatusBadRequest, middleware.NewValidationError("Project not found"))
			return
		}
	}

	id := uuid.New().String()

	var email, company, designation interface{}
	if req.Email != nil {
		email = *req.Email
		if email == "" {
			email = nil
		}
	}
	if req.Company != nil {
		company = *req.Company
		if company == "" {
			company = nil
		}
	}
	if req.Designation != nil {
		designation = *req.Designation
		if designation == "" {
			designation = nil
		}
	}

	_, err := database.Exec(
		"INSERT INTO people (id, name, email, company, designation, project_id) VALUES (?, ?, ?, ?, ?, ?)",
		id,
		req.Name,
		email,
		company,
		designation,
		req.ProjectID,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, middleware.NewCreateError("person"))
		return
	}

	var p Person
	err = database.QueryRow(
		"SELECT id, name, email, company, designation, project_id, created_at, updated_at FROM people WHERE id = ?",
		id,
	).Scan(&p.ID, &p.Name, &p.Email, &p.Company, &p.Designation, &p.ProjectID, &p.CreatedAt, &p.UpdatedAt)

	if err != nil {
		c.JSON(http.StatusInternalServerError, middleware.NewFetchError("person"))
		return
	}

	c.JSON(http.StatusCreated, middleware.NewSuccessResponse(p))
}

// UpdatePerson handles PUT /api/people/:id - Update person
func UpdatePerson(c *gin.Context) {
	databaseIface, exists := c.Get("database")
	if !exists {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Database not available"))
		return
	}

	database, ok := databaseIface.(*db.Database)
	if !ok {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Invalid database instance"))
		return
	}

	id := c.Param("id")

	// Check if person exists
	var existsCheck int
	err := database.QueryRow("SELECT COUNT(*) FROM people WHERE id = ?", id).Scan(&existsCheck)
	if err != nil || existsCheck == 0 {
		c.JSON(http.StatusNotFound, middleware.NewNotFoundError("Person"))
		return
	}

	var req UpdatePersonRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, middleware.NewValidationError("Invalid request body"))
		return
	}

	// Validate project_id if provided
	if req.ProjectID != nil {
		var count int
		err := database.QueryRow("SELECT COUNT(*) FROM projects WHERE id = ?", *req.ProjectID).Scan(&count)
		if err != nil || count == 0 {
			c.JSON(http.StatusBadRequest, middleware.NewValidationError("Project not found"))
			return
		}
	}

	// Build dynamic update query
	query := "UPDATE people SET "
	var args []interface{}
	hasUpdates := false

	if req.Name != nil {
		trimmedName := strings.TrimSpace(*req.Name)
		if trimmedName == "" {
			c.JSON(http.StatusBadRequest, middleware.NewValidationError("Person name cannot be blank"))
			return
		}
		query += "name = ?"
		args = append(args, trimmedName)
		hasUpdates = true
	}

	if req.Email != nil {
		if hasUpdates {
			query += ", "
		}
		query += "email = ?"
		if *req.Email == "" {
			args = append(args, nil)
		} else {
			args = append(args, *req.Email)
		}
		hasUpdates = true
	}

	if req.Company != nil {
		if hasUpdates {
			query += ", "
		}
		query += "company = ?"
		if *req.Company == "" {
			args = append(args, nil)
		} else {
			args = append(args, *req.Company)
		}
		hasUpdates = true
	}

	if req.Designation != nil {
		if hasUpdates {
			query += ", "
		}
		query += "designation = ?"
		if *req.Designation == "" {
			args = append(args, nil)
		} else {
			args = append(args, *req.Designation)
		}
		hasUpdates = true
	}

	if req.ProjectID != nil {
		if hasUpdates {
			query += ", "
		}
		query += "project_id = ?"
		args = append(args, *req.ProjectID)
		hasUpdates = true
	}

	if hasUpdates {
		query += ", updated_at = CURRENT_TIMESTAMP WHERE id = ?"
		args = append(args, id)
		_, err = database.Exec(query, args...)
		if err != nil {
			c.JSON(http.StatusInternalServerError, middleware.NewUpdateError("person"))
			return
		}
	}

	var p Person
	err = database.QueryRow(
		"SELECT id, name, email, company, designation, project_id, created_at, updated_at FROM people WHERE id = ?",
		id,
	).Scan(&p.ID, &p.Name, &p.Email, &p.Company, &p.Designation, &p.ProjectID, &p.CreatedAt, &p.UpdatedAt)

	if err != nil {
		c.JSON(http.StatusInternalServerError, middleware.NewFetchError("person"))
		return
	}

	c.JSON(http.StatusOK, middleware.NewSuccessResponse(p))
}

// DeletePerson handles DELETE /api/people/:id - Delete person
func DeletePerson(c *gin.Context) {
	databaseIface, exists := c.Get("database")
	if !exists {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Database not available"))
		return
	}

	database, ok := databaseIface.(*db.Database)
	if !ok {
		c.JSON(http.StatusInternalServerError, middleware.NewInternalError("Invalid database instance"))
		return
	}

	id := c.Param("id")

	// Check if person exists
	var existsCheck int
	err := database.QueryRow("SELECT COUNT(*) FROM people WHERE id = ?", id).Scan(&existsCheck)
	if err != nil || existsCheck == 0 {
		c.JSON(http.StatusNotFound, middleware.NewNotFoundError("Person"))
		return
	}

	_, err = database.Exec("DELETE FROM people WHERE id = ?", id)
	if err != nil {
		c.JSON(http.StatusInternalServerError, middleware.NewDeleteError("person"))
		return
	}

	c.JSON(http.StatusOK, middleware.NewSuccessResponse(gin.H{"message": "Person deleted successfully"}))
}
